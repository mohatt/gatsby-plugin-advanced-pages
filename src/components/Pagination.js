import React from 'react'
import { shape, number, string, object, bool, element, oneOf, oneOfType } from 'prop-types'
import classNames from 'classnames'
import Link from './Link'

export default class Pagination extends React.Component {
  static propTypes = {
    route: string.isRequired,
    params: object,
    ui: oneOf(['mini', 'simple', 'full']),
    range: number,
    className: string,
    pageInfo: shape({
      itemCount: number.isRequired,
      perPage: number.isRequired,
      pageCount: number.isRequired,
      currentPage: number.isRequired,
      hasNextPage: bool.isRequired,
      hasPreviousPage: bool.isRequired
    }).isRequired,
    labels: shape({
      prev: oneOfType([string, element]),
      next: oneOfType([string, element]),
      first: oneOfType([string, element]),
      last: oneOfType([string, element])
    }),
    theme: shape({
      inner: string,
      item: string,
      'item.next': string,
      'item.prev': string,
      'item.first': string,
      'item.last': string,
      link: string,
      active: string,
      disabled: string
    }),
    renderDisabled: bool
  };

  static defaultProps = {
    ui: 'full',
    range: 6,
    labels: {
      first: '« First',
      prev: '← Previous',
      next: 'Next →',
      last: 'Last »'
    },
    theme: {
      inner: 'pagination',
      item: 'page-item',
      'item.next': 'page-item__next',
      'item.prev': 'page-item__prev',
      'item.first': 'page-item__first',
      'item.last': 'page-item__last',
      link: 'page-link',
      active: 'active',
      disabled: 'disabled'
    },
    renderDisabled: true
  }

  render () {
    const {
      ui,
      pageInfo: {
        pageCount,
        currentPage,
        hasNextPage,
        hasPreviousPage
      }
    } = this.props

    const labels = Object.assign({}, Pagination.defaultProps.labels, this.props.labels)
    const theme = Object.assign({}, Pagination.defaultProps.theme, this.props.theme)
    const pages = []

    if (ui === 'full') {
      pages.push({
        key: 'first',
        number: 1,
        type: 'first',
        label: labels.first,
        disabled: !hasPreviousPage
      })
    }

    pages.push({
      key: 'prev',
      number: currentPage - 1,
      type: 'prev',
      label: labels.prev,
      disabled: !hasPreviousPage
    })

    if (ui !== 'mini') {
      const [fp, lp] = this.calcRange()
      for (let i = fp; i <= lp; i++) {
        pages.push({
          key: i,
          number: i,
          type: 'page',
          label: i.toString(),
          active: i === currentPage
        })
      }
    }

    pages.push({
      key: 'next',
      number: currentPage + 1,
      type: 'next',
      label: labels.next,
      disabled: !hasNextPage
    })

    if (ui === 'full') {
      pages.push({
        key: 'last',
        number: pageCount,
        type: 'last',
        label: labels.last,
        disabled: !hasNextPage
      })
    }

    return (
      <nav className={this.props.className} role='navigation' aria-label='Pagination Navigation'>
        <ul className={theme.inner}>
          {pages.map((page) => {
            if (page.disabled && !this.props.renderDisabled) {
              return null
            }

            const classes = [theme.item]
            const typeClass = theme['item.' + page.type]
            if (typeClass) {
              classes.push(typeClass)
            }

            classes.push({
              [theme.active]: page.active,
              [theme.disabled]: page.disabled
            })

            if (page.disabled) {
              return (
                <li key={page.key} className={classNames(classes)}>
                  <a className={theme.link}>{page.label}</a>
                </li>
              )
            }

            // Use non-paginated route for first page
            const [params, scope] = page.number === 1
              ? [this.props.params, null]
              : [{ ...this.props.params, page: page.number }, 'pagination']

            return (
              <li key={page.key} className={classNames(classes)}>
                <Link to={this.props.route} params={params} scope={scope} className={theme.link}>
                  {page.label}
                </Link>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

  calcRange () {
    const { range, pageInfo: { currentPage, pageCount } } = this.props

    let fp = Math.max(1, currentPage - Math.floor(range / 2))
    let lp = Math.min(pageCount, currentPage + Math.floor(range / 2))

    if (lp - fp + 1 < range) {
      if (currentPage < pageCount / 2) {
        lp = Math.min(
          pageCount,
          lp + (range - (lp - fp))
        )
      } else {
        fp = Math.max(1, fp - (range - (lp - fp)))
      }
    }

    if (lp - fp + 1 > range) {
      if (currentPage > pageCount / 2) {
        fp = fp + 1
      } else {
        lp = lp - 1
      }
    }

    return [fp, lp]
  }
}
