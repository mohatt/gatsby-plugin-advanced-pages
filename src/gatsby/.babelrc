{
  "presets": [
    ["babel-preset-gatsby-package", { "nodeVersion": "8" }]
  ]
}

