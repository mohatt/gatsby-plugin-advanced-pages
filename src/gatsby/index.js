export { default as onPreBootstrap } from './bootstrap'
export { default as sourceNodes } from './source-nodes'
export { default as setFieldsOnGraphQLNodeType } from './extend-nodes'
export { default as onCreateNode } from './on-create-node'
export { default as createPages } from './create-pages'
export { default as onPreExtractQueries } from './install-fragments'
