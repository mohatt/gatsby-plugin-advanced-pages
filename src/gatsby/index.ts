export type * from './types'
export type { RouteParams, PathGenerator } from '../lib/route-compiler'
