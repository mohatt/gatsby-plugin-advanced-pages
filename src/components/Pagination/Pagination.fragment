import { graphql } from "gatsby"

export const pagination = graphql`
  fragment Pagination on PageInfo {
    perPage
    pageCount
    itemCount
    currentPage
    hasNextPage
    hasPreviousPage
  }
`
