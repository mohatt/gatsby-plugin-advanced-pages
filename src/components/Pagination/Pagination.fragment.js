import { graphql } from 'gatsby'

export const Pagination = graphql`
  fragment Pagination on PageInfo {
    perPage
    pageCount
    itemCount
    currentPage
    hasNextPage
    hasPreviousPage
  }
`
