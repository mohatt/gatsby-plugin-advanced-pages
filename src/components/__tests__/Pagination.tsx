// @vitest-environment jsdom
import React from 'react'
import ReactTest from '@testing-library/react'
import _ from 'lodash'
import { graphql } from 'gatsby'
import { expect, Mock } from 'vitest'
import type { DeepPartial } from '@/types'
import defaultTestCase, { testCases } from './__fixtures__/pagination'
import { Pagination, PaginationProps } from '../Pagination'
import { Pagination as PaginationFragment } from '../Pagination.fragment'

vi.mock('../Link', () => ({
  Link: (props: any) => <a>{JSON.stringify(props)}</a>,
}))

describe('<Pagination />', () => {
  const render = (props?: DeepPartial<PaginationProps>) => {
    const fullProps = _.merge({}, defaultTestCase, props)
    return ReactTest.render(<Pagination {...fullProps} />).container.firstChild
  }

  it('renders correctly with defaults', () => {
    expect(render()).toMatchSnapshot()
  })

  it('renders correctly with custom theme', () => {
    expect(
      render({
        theme: _.mapValues(Pagination.defaultProps.theme, (v) => v + '-custom'),
      }),
    ).toMatchSnapshot()
  })

  it('renders correctly with custom labels', () => {
    expect(
      render({
        labels: _.mapValues(Pagination.defaultProps.labels, (v) => v + ' custom'),
      }),
    ).toMatchSnapshot()
  })

  it('renders correctly with mini ui', () => {
    expect(render({ ui: 'mini' })).toMatchSnapshot()
  })

  it('renders correctly with simple ui', () => {
    expect(render({ ui: 'simple' })).toMatchSnapshot()
  })

  it('renders correctly with full ui', () => {
    expect(render({ ui: 'full' })).toMatchSnapshot()
  })

  it('renders correctly with custom range', () => {
    expect(render({ range: 1 })).toMatchSnapshot()
    expect(render({ range: 3 })).toMatchSnapshot()
    expect(render({ range: 7 })).toMatchSnapshot()
    expect(render({ range: 9 })).toMatchSnapshot()
  })

  it("shouldn't render disabled nav items when renderDisabled is false", () => {
    expect(
      render({
        pageInfo: { currentPage: 1, hasPreviousPage: false },
        renderDisabled: false,
      }),
    ).toMatchSnapshot()
    expect(
      render({
        pageInfo: { currentPage: defaultTestCase.pageInfo.pageCount, hasNextPage: false },
        renderDisabled: false,
      }),
    ).toMatchSnapshot()
    expect(
      render({
        pageInfo: {
          itemCount: 1,
          perPage: 10,
          pageCount: 1,
          currentPage: 1,
          hasNextPage: false,
          hasPreviousPage: false,
        },
        renderDisabled: false,
      }),
    ).toMatchSnapshot()
  })

  it.each(testCases)('correctly handles pageInfo options ($id)', ({ props }) => {
    expect(render(props)).toMatchSnapshot()
  })

  it('correctly defines pagination fragment', () => {
    expect(PaginationFragment).toBeUndefined()
    expect((graphql as Mock).mock.calls).toMatchSnapshot()
  })
})
