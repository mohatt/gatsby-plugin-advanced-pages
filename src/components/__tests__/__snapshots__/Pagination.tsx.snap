// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Pagination /> > correctly defines pagination fragment 1`] = `
[
  [
    [
      "
  fragment Pagination on PageInfo {
    perPage
    pageCount
    itemCount
    currentPage
    hasNextPage
    hasPreviousPage
  }
",
    ],
  ],
]
`;

exports[`<Pagination /> > correctly handles pageInfo options ('1-page-first') 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first disabled"
    >
      <a
        class="page-link"
      >
        « First
      </a>
    </li>
    <li
      class="page-item page-item__prev disabled"
    >
      <a
        class="page-link"
      >
        ← Previous
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item page-item__next disabled"
    >
      <a
        class="page-link"
      >
        Next →
      </a>
    </li>
    <li
      class="page-item page-item__last disabled"
    >
      <a
        class="page-link"
      >
        Last »
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > correctly handles pageInfo options ('3-page-last') 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item page-item__next disabled"
    >
      <a
        class="page-link"
      >
        Next →
      </a>
    </li>
    <li
      class="page-item page-item__last disabled"
    >
      <a
        class="page-link"
      >
        Last »
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > correctly handles pageInfo options ('12-page-8th') 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":7},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":7},"scope":"pagination","className":"page-link","children":"7"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":8},"scope":"pagination","className":"page-link","children":"8"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"9"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":10},"scope":"pagination","className":"page-link","children":"10"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":11},"scope":"pagination","className":"page-link","children":"11"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":12},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with custom labels 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First custom"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous custom"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next → custom"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last » custom"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with custom range 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with custom range 2`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with custom range 3`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":7},"scope":"pagination","className":"page-link","children":"7"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with custom range 4`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":7},"scope":"pagination","className":"page-link","children":"7"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":8},"scope":"pagination","className":"page-link","children":"8"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"9"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with custom theme 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination-custom"
  >
    <li
      class="page-item-custom page-item__first-custom"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link-custom","children":"« First"}
      </a>
    </li>
    <li
      class="page-item-custom page-item__prev-custom"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link-custom","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item-custom"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link-custom","children":"1"}
      </a>
    </li>
    <li
      class="page-item-custom"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link-custom","children":"2"}
      </a>
    </li>
    <li
      class="page-item-custom"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link-custom","children":"3"}
      </a>
    </li>
    <li
      class="page-item-custom active-custom"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link-custom","children":"4"}
      </a>
    </li>
    <li
      class="page-item-custom"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link-custom","children":"5"}
      </a>
    </li>
    <li
      class="page-item-custom"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link-custom","children":"6"}
      </a>
    </li>
    <li
      class="page-item-custom page-item__next-custom"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link-custom","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item-custom page-item__last-custom"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link-custom","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with defaults 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with full ui 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with mini ui 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > renders correctly with simple ui 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > shouldn't render disabled nav items when renderDisabled is false 1`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"2"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":3},"scope":"pagination","className":"page-link","children":"3"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item page-item__next"
    >
      <a>
        {"to":"whatever","params":{"page":2},"scope":"pagination","className":"page-link","children":"Next →"}
      </a>
    </li>
    <li
      class="page-item page-item__last"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"Last »"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > shouldn't render disabled nav items when renderDisabled is false 2`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item page-item__first"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"« First"}
      </a>
    </li>
    <li
      class="page-item page-item__prev"
    >
      <a>
        {"to":"whatever","params":{"page":8},"scope":"pagination","className":"page-link","children":"← Previous"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":4},"scope":"pagination","className":"page-link","children":"4"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":5},"scope":"pagination","className":"page-link","children":"5"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":6},"scope":"pagination","className":"page-link","children":"6"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":7},"scope":"pagination","className":"page-link","children":"7"}
      </a>
    </li>
    <li
      class="page-item"
    >
      <a>
        {"to":"whatever","params":{"page":8},"scope":"pagination","className":"page-link","children":"8"}
      </a>
    </li>
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","params":{"page":9},"scope":"pagination","className":"page-link","children":"9"}
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> > shouldn't render disabled nav items when renderDisabled is false 3`] = `
<nav
  aria-label="Pagination Navigation"
  role="navigation"
>
  <ul
    class="pagination"
  >
    <li
      class="page-item active"
    >
      <a>
        {"to":"whatever","scope":null,"className":"page-link","children":"1"}
      </a>
    </li>
  </ul>
</nav>
`;
